lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,197 @@
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const STORE_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  getRemoteSettings(name) {
+    return RemoteSettings(name).get();
+  }
+
+  getRecipeExecutor(nbTaggers, nmfTaggers) {
+    return new RecipeExecutor(nbTaggers, nmfTaggers);
+  }
+
+  getNaiveBayesTextTagger(model) {
+    return new NaiveBayesTextTagger(model);
+  }
+
+  getNmfTextTagger(model) {
+    return new NmfTextTagger(model);
+  }
+
+  getNewTabUtils() {
+    return NewTabUtils;
+  }
+
+  /**
+   * Returns a Recipe from remote settings to be consumed by a RecipeExecutor.
+   * A Recipe is a set of instructions on how to processes a RecipeExecutor.
+   */
+  async getRecipe() {
+    if (!this.recipe) {
+      this.recipe = await this.getRemoteSettings("personality-provider-recipe");
+    }
+    return this.recipe[0];
+  }
+
+  /**
+   * Returns a Recipe Executor.
+   * A Recipe Executor is a set of actions that can be consumed by a Recipe.
+   * The Recipe determines the order and specifics of which the actions are called.
+   */
+  async generateRecipeExecutor() {
+    let nbTaggers = [];
+    let nmfTaggers = {};
+    const models = await this.getRemoteSettings("personality-provider-models");
```

Comment:
Again, `models` could be an empty array. Shall we short-circuit, or create a `RecipeExecutor` regardless?

Reply:
short-circuit?

Reply:
In case of `getRemoteSettings` returns an empty array (we've seen this happened in prod), shall we just bail out instead of generating a malformed recipe executor?

Reply:
Ah, I'm not sure. If we bail, I think things are going to be pretty wonky. I think the safest thing to do is return empty arrays, which if it doesn't already, I can ensure causes everything to do nothing.

It would end up generating something as if it was a fresh profile with no history.

Reply:
Yep, I was wondering that if we could just abort this round completely until the Remote Settings is ready. @jonathankoren what do you think?

This also reminds me that shall we introduce some mechanism to trigger this whole personalization. Like if RS is not ready or Places doesn't return enough records, then just wait for next cycle.

Reply:
Hm, what does it mean when you say "RS is not ready"? Does it go through a state change where, it's not ready, and using it is slow, and it is ready and using it is fast, or something?

Reply:
In short, `RS.get()` might run into some trouble in the wild. For example, this is what we've observed in production "NetworkError when attempting to fetch resource". It printed this error message and returned an empty array, perhaps would retry it later.